Undefined reference to hpx::assertion_failed

Building commit 6c08935c5d9158031bbdee01872283e2dd8b0ceb with GCC 5.3 and RelWithDebInfo as the build type fails due to an undefined reference to `hpx::assertion_failed`.

After examining the compiler output, it seems that all calls of assertion_failed are inlined. Afterwards the dead function is eliminated by the compiler and the reference in assert.hpp can not subsequently be resolved by the linker.

The reason for this behavior seems to be the contradicting declaration of this function in exception.hpp (as an inline function) and in assert.hpp (as a non-inline function).

As far as I know an inline function must be defined in every translation unit where it is called. In addition, an inline function with external linkage must be declared inline in every translation unit. Both rules are (possibly) violated in this case.

I was able to fix the issue by moving the definition into exception.cpp

assertion_failed_msg should be affected by the same issue.

--- a/hpx/exception.hpp
+++ b/hpx/exception.hpp
@@ -1554,11 +1554,13 @@ namespace hpx
     }
 
         // forwarder for HPX_ASSERT handler
-    HPX_EXPORT void assertion_failed(char const* expr, char const* function,
+    HPX_ATTRIBUTE_NORETURN HPX_EXPORT
+    void assertion_failed(char const* expr, char const* function,
         char const* file, long line);
 
     // forwarder for HPX_ASSERT_MSG handler
-    HPX_EXPORT void assertion_failed_msg(char const* msg, char const* expr,
+    HPX_ATTRIBUTE_NORETURN HPX_EXPORT
+    void assertion_failed_msg(char const* msg, char const* expr,
         char const* function, char const* file, long line);
 
     // For testing purposes we sometime expect to see exceptions, allow those

--- a/hpx/exception.hpp
+++ b/hpx/exception.hpp
@@ -1554,18 +1554,12 @@ namespace hpx
     }
 
         // forwarder for HPX_ASSERT handler
-    inline void assertion_failed(char const* expr, char const* function,
-        char const* file, long line)
-    {
-        hpx::detail::assertion_failed(expr, function, file, line);
-    }
+    HPX_EXPORT void assertion_failed(char const* expr, char const* function,
+        char const* file, long line);
 
     // forwarder for HPX_ASSERT_MSG handler
-    inline void assertion_failed_msg(char const* msg, char const* expr,
-        char const* function, char const* file, long line)
-    {
-        hpx::detail::assertion_failed_msg(msg, expr, function, file, line);
-    }
+    HPX_EXPORT void assertion_failed_msg(char const* msg, char const* expr,
+        char const* function, char const* file, long line);
 
     // For testing purposes we sometime expect to see exceptions, allow those
     // to go through without attaching a debugger.

--- a/src/exception.cpp
+++ b/src/exception.cpp
@@ -1137,6 +1137,18 @@ namespace hpx
     {
         return get_error_state(detail::access_exception(e));
     }
+
+    void assertion_failed(char const* expr, char const* function,
+        char const* file, long line)
+    {
+        hpx::detail::assertion_failed(expr, function, file, line);
+    }
+
+    void assertion_failed_msg(char const* msg, char const* expr,
+        char const* function, char const* file, long line)
+    {
+        hpx::detail::assertion_failed_msg(msg, expr, function, file, line);
+    }
 }
 
 

